AppController/djinn.rb

```diff
@@ -2166,6 +2170,129 @@ def job_start(secret)
+      rescue Errno::ENOENT
+        # On appscale down, terminate may delete our secret key before we
+        # can check it here.
+        Djinn.log_debug("run_terminate(): didn't find secret file. Continuing.")
+      end
+      Djinn.log_info("Received a stop request.")
+      Djinn.log_info("Stopping all other nodes.")
+      terminate_appscale_in_parallel(clean){|node| send_client_message(node)}
+      @done_terminating = true
+    end
+  end
+
+
+  def terminate_appscale(my_node, clean)
+    ip = my_node.private_ip
+    Djinn.log_info("Running terminate.rb on the node at IP address #{ip}")
+    ssh_key = my_node.ssh_key
+
+    Djinn.log_info("clean val: #{clean.downcase}")
+    # Add ' clean' as parameter to terminate.rb if clean is true
+    extra_command = clean.downcase == 'true' ? ' clean' : ''
+
+    # Run terminate.rb on node
+    HelperFunctions.sleep_until_port_is_open(ip, SSH_PORT)
+    output = HelperFunctions.run_remote_command(ip,
+        "ruby /root/appscale/AppController/terminate.rb#{extra_command}",
+        ssh_key, true)
+
+    # terminate.rb will print "OK" if it ran successfully
+    status = output.chomp!("OK").nil? ? false : true
+
+    # Get the output of 'ps x' from node
+    output += HelperFunctions.run_remote_command(ip, 'ps x', ssh_key, true)
+    Djinn.log_info("#{ip} terminated:#{status}\noutput:#{output}")
+
+    # Return ip, status, and output of terminated node
+    return {'ip'=>ip, 'status'=> status, 'output'=>output}
+  end
+
+
+  def terminate_appscale_in_parallel(clean)
+    # Let's stop all other nodes.
+    threads = []
+    @nodes.each { |node|
+      if node.private_ip != my_node.private_ip
+        threads << Thread.new {
+          Thread.current[:output] = terminate_appscale(node, clean)
+        }
+      end
+    }
+
+    threads.each do |t|
+      t.join
+      yield t[:output]
+    end
+
+    return "OK"
+  end
+
+
+  # Adds message to queue so it can be received by client.
+  def send_client_message(message)
+    @waiting_messages.synchronize {
+      @waiting_messages.push(message)
+      Djinn.log_info(@waiting_messages)
+      @message_ready.signal
+    }
+  end
+
+  # Method ran by client to access the messages in @waiting_messages. Client
+  # must send a timeout so that the server will not clear the messages.
+  def receive_server_message(timeout, secret)
+    was_queue_emptied = false
+    begin
+      Timeout.timeout(timeout.to_i) {
+        begin
+          bad_secret = JSON.dump({'status'=>BAD_SECRET_MSG})
+          return bad_secret unless valid_secret?(secret)
+        rescue Errno::ENOENT
+          # On appscale down, terminate may delete our secret key before we
+          # can check it here.
+          Djinn.log_debug("run_terminate(): didn't find secret file. Continuing.")
+        end
+        # Client will process "Error" and try again unless appscale is
+        # terminated
+        if @done_terminating and @waiting_messages.empty?
+          return "Error"
```

Comment:
Can we replace this with something else? for example, AppScale is still running? or Not dead yet?